Namespace API Lock docs (#13064)
hashicorp/vault: website/content/docs/concepts/namespace-api-lock.mdx

---
layout: docs
page_title: Namespace API Lock
description: Lock the Vault API on a per-namespace basis.
---

> # Lock/Unlock

Reviewer: ```suggestion
# Namespace Lock and Unlock
```

Vault makes the API available (unlocked) by default for all namespaces. In
this state, Vault can respond to all API/CLI ('API' from here on out) requests
as it usually would.

A Vault Administrator can lock the API for particular namespaces. In this state,
Vault will block all but a select few API endpoints from responding to clients
operating in a locked namespace (or a descendant of a locked namespace). The
endpoints that do respond, the exempt paths, are largely the same as the Vault
unauthenticated paths. They are mainly used for checking the status of various
systems (e.g. `sys/health`), or for unlocking the API.

When the API is locked for a particular namespace, it is also locked for all

> descendants of that namespace. If the API was already locked for a descendant

Reviewer: ```suggestion
descendants of that namespace. If the API was already locked for a descendant,
```

that lock will remain, but Vault must be unlocked for the parent before
unlocking the descendant.

> ## Why?

Reviewer: Maybe **Benefits of Namespace Lock** is a better heading?

> Blocking access to much of Vault can be an important break-glass tool in the

Reviewer: ```suggestion
Blocking access to much of Vault can be advantageous in the
```

> event of unexpected behavior.

Reviewer: If there are specific examples of these unexpected behavior, it might be good to call them out.

> For HCP Vault, this provides functionality analogous to sealing Vault, without

Reviewer: This transit to HCP Vault here is a bit confusing to me. 

Unless you are trying to make that analogy between seal/unseal and namespace for Vault. If that is the intent, then I would just start out by saying: The namespace lock/unlock functionality is analogous to the seal/unseal feature, but without requiring the Vault administrator to request the Managed Service Provider to seal/unseal Vault.

Something along those lines?

the Vault Administrator requesting that the Managed Service Provider seal/unseal
Vault.

This feature also becomes valuable for secure multitenancy in a variety of Vault

> deployment strategies. One can restrict Vault access for just part of an

Reviewer: ```suggestion
deployment strategies. You can restrict Vault access for just part of an
```

> organization, without affecting adjacent parts of the business. If unexpected

Reviewer: ```suggestion
organization without affecting adjacent parts of the business. For example, if unexpected
```

> behavior is detected in sub-organization A, an Administrator can disable Vault

Reviewer: ```suggestion
behavior is detected in sub-organization A, an administrator can disable Vault
```

> access for any entity under A, without disabling access for sub-organization B.

Reviewer: ```suggestion
access for any entity under sub-organization A without disabling access for sub-organization B.
```

> access for any entity under A, without disabling access for sub-organization B.

Reviewer: ```suggestion
access for any entity under sub-organization A, without disabling access for sub-organization B.
```

> Once the cause has been identified and resolved, the API can be unlocked for A.

Reviewer: ```suggestion
Once the issue has been identified and resolved, the API can be unlocked for sub-organization A.
```

## Locking

The API can be locked by running `vault namespace lock` (or via the API) while
operating in the namespace to lock. Optionally, a subpath can be provided to
lock a descendant of the current namespace.

An unlock key will be returned, which can be used to unlock the API for that
namespace. Preserve this key to unlock the API in the future.

When the API is locked for a namespace, it will also be locked for all
descendants of that namespace. If an authenticated client attempts to access
Vault from a locked namespace, the returned error will inform that client of the

> locking namespace.

Reviewer: Suggested text:

If an authenticated client attempts to access Vault from a locked namespace, the return error informs the client that they are attempting to access a locked namespace.

## Unlocking

The API can be unlocked by running `vault namespace unlock` (or via the API)

> while operating in the namespace to unlock. Optionally a subpath can be

Reviewer: ```suggestion
while operating in the namespace to unlock. Optionally, a subpath can be
```

provided to lock a descendant of the current namespace.

In general, an unlock key is required to unlock the API. This is the same as the
unlock key provided when the namespace was locked.

The unlock key requirement can be overriden by using a root token with the
unlock request. In this case, the unlock key does not need to be provided.

When the API is unlocked, it will also be unlocked for all descendants that were
locked with it. If a descendant namespace was previously locked, that lock will

> remain.

Reviewer: ```suggestion
remain in place.
```

## API Locked Response

If a request is made to a non-exempt path from a locked namespace, e.g. `nsA`,

> Vault will respond with an HTTP 503: Service Unavailable. It will also yield an

Reviewer: ```suggestion
Vault responds with an HTTP 503: Service Unavailable. It will also produce the following error:
```

> error of the form:

Reviewer: ```suggestion

```

`API access to this namespace has been locked by an administrator - "nsA" must be unlocked to gain access.`

> Similarly, the same error will be returned if a request is made in a descendant

Reviewer: ```suggestion
Similarly, the same error will return if a request is made in a descendant
```
